tests/testthat/test-return_linter.R

```diff
@@ -0,0 +1,841 @@
+test_that("Lint return on end of function", {
+  expect_lint(
+    trim_some("
+      function() {
+        return(1)
+        # Test
+        3 + 4
+      }
+    "),
+    list(
+      line_number = 4L,
+      message = rex::rex("All functions must have an explicit return().")
+    ),
+    return_linter(use_implicit_returns = FALSE)
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        return(1)
+      }
+    "),
+    list(
+      line_number = 2L,
+      message = rex::rex("Use implicit return behavior; explicit return() is not needed.")
+    ),
+    return_linter()
+  )
+})
+
+test_that("Lint return on end of lambda function", {
+  skip_if_not_r_version("4.1.0")
+
+  expect_lint(
+    trim_some("
+      \\(bar) {
+        5L + 3L
+      }
+    "),
+    list(
+      line_number = 2L,
+      message = rex::rex("All functions must have an explicit return().")
+    ),
+    return_linter(use_implicit_returns = FALSE)
+  )
+
+  expect_lint(
+    trim_some("
+      \\(bar) {
+        5L + 3L
+        return(1)
+      }
+    "),
+    list(
+      line_number = 3L,
+      message = rex::rex("Use implicit return behavior; explicit return() is not needed.")
+    ),
+    return_linter()
+  )
+})
+
+test_that("Do not lint control statements (with return) on end of function", {
+  linter <- return_linter(use_implicit_returns = FALSE)
+
+  expect_lint(
+    trim_some("
+      function() {
+        repeat {
+          cat(4)
+        }
+      }
+    "),
+    NULL,
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        if(x) {
+          return(4)
+        } else if(y) {
+          return(5)
+        } else {
+          return(6)
+        }
+      }
+    "),
+    NULL,
+    linter
+  )
+})
+
+test_that("Lint control statements (without return) on end of function", {
+  linter <- return_linter(use_implicit_returns = FALSE)
+  msg <- rex::rex("All functions must have an explicit return().")
+
+  expect_lint(
+    trim_some("
+      function() {
+        while(x > 4) {
+          cat(4)
+          if(x < 4) {
+            return(x)
+          }
+        }
+      }
+    "),
+    list(msg, line_number = 2L),
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        for(i in 1:10) {
+          cat(4)
+          if(i > 11) {
+            return(x)
+          }
+        }
+      }
+    "),
+    list(msg, line_number = 2L),
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        if(x == 2L){
+          return(e)
+        }
+      }
+    "),
+    list(msg, line_number = 2L),
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      function() {
+        if(x == 2L){
```

Comment:
This code should be reformatted to have some whitespace around ().

Reply:
Thanks... I would hold off on stylistic comments until the PR is marked for review (not draft). Some of this will be removed gradually as tests are fixed.